crates/table/src/table.rs

```diff
@@ -234,20 +250,242 @@ impl Table {
+                    // - The `fixed_buf` is exactly the right `fixed_row_size`.
+                    unsafe { static_layout.deserialize_row_into(fixed_buf, row) };
+                    Ok(fixed_offset)
+                })
+                .map_err(Error::PagesError)?;
+            match res {
+                (page, Ok(offset)) => (RowPointer::new(false, page, offset, squashed_offset), 0.into()),
+                (_, Err(e)) => return Err(e),
+            }
+        } else {
+            // SAFETY: `self.pages` is known to be specialized for `self.row_layout`,
+            // as `self.pages` was constructed from `self.row_layout` in `Table::new`.
+            unsafe {
+                write_row_to_pages_bsatn(
+                    &mut self.inner.pages,
+                    &self.inner.visitor_prog,
+                    blob_store,
+                    &self.inner.row_layout,
+                    row,
+                    self.squashed_offset,
+                )
+            }?
+        };
+
+        // SAFETY: We just inserted `ptr`, so it must be present.
+        let row_ref = unsafe { self.inner.get_row_ref_unchecked(blob_store, ptr) };
+
+        Ok((row_ref, blob_bytes))
+    }
+
+    /// Returns all the columns with sequences that need generation for this `row`.
+    ///
+    /// # Safety
+    ///
+    /// `self.is_row_present(row)` must hold.
+    pub unsafe fn sequence_triggers_for<'a>(
+        &'a self,
+        blob_store: &'a dyn BlobStore,
+        row: RowPointer,
+    ) -> (ColList, SeqIdList) {
+        let sequences = &*self.get_schema().sequences;
+        let row_ty = self.row_layout().product();
+
+        // SAFETY: Caller promised that `self.is_row_present(row)` holds.
+        let row_ref = unsafe { self.get_row_ref_unchecked(blob_store, row) };
+
+        sequences
+            .iter()
+            // Find all the sequences that are triggered by this row.
+            .filter(|seq| {
+                // SAFETY: `seq.col_pos` is in-bounds of `row_ty.elements`
+                // as `row_ty` was derived from the same schema as `seq` is part of.
+                let elem_ty = unsafe { &row_ty.elements.get_unchecked(seq.col_pos.idx()) };
+                // SAFETY:
+                // - `elem_ty` appears as a column in th row type.
+                // - `AlgebraicValue` is compatible with all types.
+                let val = unsafe { AlgebraicValue::unchecked_read_column(row_ref, elem_ty) };
+                val.is_numeric_zero()
+            })
+            .map(|seq| (seq.col_pos, seq.sequence_id))
+            .unzip()
+    }
+
+    /// Writes `seq_val` to the column at `col_id` in the row identified by `ptr`.
+    ///
+    /// Truncates the `seq_val` to fit the type of the column.
+    ///
+    /// # Safety
+    ///
+    /// - `self.is_row_present(row)` must hold.
+    /// - `col_id` must be a valid column, with a primiive type, of the row type.
```

Comment:
```suggestion
    /// - `col_id` must be a valid column, with a primitive integer type, of the row type.
```

Reply:
Spelling on "primitive." Being an integer is not a safety requirement as written, but I think it could be (see other comment). Whether or not we actually invoke UB if the column is non-integer, I think it's simplest in terms of docs to just say that we don't define behavior in that case, i.e. to make it a safety requirement as I'm doing here.